# Introduction

This repository contains processed 16S rRNA gene (V3-V4) sequencing data and analysis code related to below paper.<br/>
- Title: **Assembly and maturation of calf gut microbiome from neonate to post puberty**
- Authors: **Hyun Sik Kim, Jae-Yun Lee, Tae Woong Whon, and Jin-Woo Bae**
- Year: 2024

# Contents
## Processed data in analysis-ready format
- **Processed_data.zip**: A compressed zip file containing the processed data, ready for immediate analysis. It comprises the following files:
  - `ASV_table.tsv`: Amplicon sequencing variant (ASV) abundance table. The columns are sample names, and the rows are ASV names. Generated from [DADA2](https://benjjneb.github.io/dada2/index.html) pipeline through [Qiime2](https://qiime2.org/) platform.
  - `Representative_ASV_sequences.fasta`: Sequences of each ASV in FASTA format. Generated from [DADA2](https://benjjneb.github.io/dada2/index.html) pipeline through [Qiime2](https://qiime2.org/) platform.
  - `Phylogenetic_tree.nwk`: Newick format phylogenetic tree of ASVs, generated by [q2-fragment-insertion](https://library.qiime2.org/plugins/q2-fragment-insertion/16/) plugin in [Qiime2](https://qiime2.org/) platform.  
  - `Taxonomy.tsv`: Taxonomic annotation of ASVs. Generated from [Qiime2](https://qiime2.org/) platform.
  - `Shannon_alpha_rarefaction.csv`: Rarefaction curve of our data based on Shannon diversity index, generated from [Qiime2](https://qiime2.org/) platform.
  - `Metadata.tsv`: Metadata containing sample information available for analysis, such as sampling time points, sample origin and host sex.

## `Analysis_codes` containing two custom codes
  - `Primer_trimming.py`: Python script used for trimming 338F and 805R primer from the raw reads using [Cutadapt](https://cutadapt.readthedocs.io/en/stable/) before DADA2 pipeline.
    ```bash
    # Example usage
    chmod 755 Primer_trimming.py # Set permissions for the script
    ./Primer_trimming.py -h # For more detail
    ```
  - `Draw_figures.R`: R script used for generating the figures presented in our paper. All the data used for this script are in `Processed_data.zip`.
